scipy/optimize/_minpack_py.py

```diff
@@ -465,11 +466,24 @@ def leastsq(func, x0, args=(), Dfun=None, full_output=0,
     if full_output:
         cov_x = None
         if info in LEASTSQ_SUCCESS:
-            perm = take(eye(n), retval[1]['ipvt'] - 1, 0)
+            # This was
+            # perm = take(eye(n), retval[1]['ipvt'] - 1, 0)
+            # r = triu(transpose(retval[1]['fjac'])[:n, :])
+            # R = dot(r, perm)
+            # cov_x = inv(dot(transpose(R), R))
+            # but the explicit dot product was not necessary and sometimes
+            # the result was not symmetric positive definite. See gh-4555.
+            perm = retval[1]['ipvt'] - 1
+            n = len(perm)
             r = triu(transpose(retval[1]['fjac'])[:n, :])
-            R = dot(r, perm)
+            inv_triu = linalg.get_lapack_funcs('trtri', (r,))
             try:
-                cov_x = inv(dot(transpose(R), R))
+                # inverse of permuted matrix is a permuation of matrix inverse
+                invR, trtri_info = inv_triu(r)  # default: upper, non-unit diag
+                if trtri_info:
```

Comment:
I think it is more readable if we explicitly make it `if trtri_info != 0`.

Reply:
OK.